### PR TITLE
Enable python API for NCEPLIBS-bufr.

### PR DESCRIPTION
**Pull request by aerorahul**
This PR:

- brings in the python API `py-ncepbufr` for NCEPLIBS-bufr from @jswhit2.  This is optional and can be enabled via `-DENABLE_PYTHON=ON`
- builds the html documentation with `pdoc`.  The built html documentation is placed in the `docs/python` directory.  I don't know how to reference this from doxygen or use doxygen for python documentation.  I am sure there is a way.  Help wanted to use doxygen with python documentation or to place a link on the main documentation page.
- adds tests that exercise the methods in the python API.  The tests exercise almost all the methods implemented in the python API.  Moreover, they also provide example usage of how to read the different bufr files e.g. `prepbufr`, `gpsbufr`, `amsuabufr`, `satwndbufr`, etc.  I cannot say exactly how much of the code is covered in terms of percentage.  Help wanted to put a number on code coverage!
- enables python in CI.  The python tests will fail because the test data needs to be updated on the FTP site to include the data for python tests.

I believe I have attempted to satisfy all the requirements; issue before PR, documentation, testing, and CI.  If there is anything else missing from this PR, please let me know.

### CODE REVIEW

**Comment by edwardhartnett** [comment]
I will add links to the project documentation to what ends up in docs/python...

**Comment by jbathegit** [comment]
If you can provide the python test data, I can get the bufr.tar updated on emcrzdm (i.e. the FTP site).   Currently the file contents are as follows:

jator2@vm-lnx-emcrzdm02.ncep.noaa.gov [/home/ftp/emc/static_files/public (44)] % tar tvf bufr.tar
drwxr-sr-x Jeff.Ator/emcobs  0 2020-11-06 18:04 bufr/
drwxr-sr-x Jeff.Ator/emcobs  0 2020-11-09 17:21 bufr/testfiles/
-rw-r--r-- Jeff.Ator/emcobs 240205 2020-11-09 21:02 bufr/testfiles/OUT_5
-rw-r--r-- Jeff.Ator/emcobs   7176 2020-11-06 18:05 bufr/testfiles/OUT_4_infile1
-rw-r--r-- Jeff.Ator/emcobs 4069648 2020-11-06 18:05 bufr/testfiles/OUT_2_preAPX
-rw-r--r-- Jeff.Ator/emcobs   19936 2020-11-06 18:05 bufr/testfiles/IN_2
-rw-r--r-- Jeff.Ator/emcobs  235622 2020-11-06 18:05 bufr/testfiles/OUT_4_infile2
-rw-r--r-- Jeff.Ator/emcobs  261739 2020-11-06 18:05 bufr/testfiles/bufrtab.TableD_STD_0_14
-rw-r--r-- Jeff.Ator/emcobs   37611 2020-11-06 18:05 bufr/testfiles/IN_3
-rw-r--r-- Jeff.Ator/emcobs    3588 2020-11-06 18:05 bufr/testfiles/IN_4
-rw-r--r-- Jeff.Ator/emcobs   51444 2020-11-09 17:21 bufr/testfiles/OUT_5_infile
-rw-r--r-- Jeff.Ator/emcobs   66976 2020-11-06 18:05 bufr/testfiles/IN_5
-rw-r--r-- Jeff.Ator/emcobs  149506 2020-11-06 18:05 bufr/testfiles/bufrtab.TableB_STD_0_16
-rw-r--r-- Jeff.Ator/emcobs  281973 2020-11-06 18:05 bufr/testfiles/bufrtab.TableD_STD_0_16
-rw-r--r-- Jeff.Ator/emcobs  503945 2020-11-06 18:05 bufr/testfiles/bufrtab.CodeFlag_LOC_0_7_1
-rw-r--r-- Jeff.Ator/emcobs    4169 2020-11-06 18:05 bufr/testfiles/IN_1
-rw-r--r-- Jeff.Ator/emcobs 4069800 2020-11-06 18:05 bufr/testfiles/OUT_2
-rw-r--r-- Jeff.Ator/emcobs  129681 2020-11-06 18:05 bufr/testfiles/bufrtab.TableB_STD_0_13
-rw-r--r-- Jeff.Ator/emcobs    3968 2020-11-06 18:05 bufr/testfiles/OUT_1
-rw-r--r-- Jeff.Ator/emcobs   11259 2020-11-06 18:05 bufr/testfiles/OUT_3_bufrtab
-rw-r--r-- Jeff.Ator/emcobs  143690 2020-11-06 18:05 bufr/testfiles/bufrtab.TableB_STD_0_14
-rw-r--r-- Jeff.Ator/emcobs  210276 2020-11-06 18:05 bufr/testfiles/bufrtab.TableD_STD_0_13
-rw-r--r-- Jeff.Ator/emcobs  241392 2020-11-06 18:05 bufr/testfiles/OUT_4
-rw-r--r-- Jeff.Ator/emcobs    6352 2020-11-06 18:05 bufr/testfiles/OUT_3
-rw-r--r-- Jeff.Ator/emcobs   38475 2020-11-06 18:05 bufr/testfiles/IN_2_bufrtab
-rw-r--r-- Jeff.Ator/emcobs   59719 2020-11-06 18:05 bufr/testfiles/bufrtab.TableB_LOC_0_7_1
-rw-r--r-- Jeff.Ator/emcobs    1463 2020-11-06 18:05 bufr/testfiles/bufrtab.TableD_LOC_0_7_1
-rw-r--r-- Jeff.Ator/emcobs    7047 2020-11-06 18:05 bufr/testfiles/OUT_1_bufrtab
-rw-r--r-- Jeff.Ator/emcobs  282554 2020-11-06 18:05 bufr/testfiles/bufrtab.CodeFlag_STD_0_13
-rw-r--r-- Jeff.Ator/emcobs   53345 2020-11-06 18:05 bufr/testfiles/OUT_2_bufrtab

So we could stash the python test data anywhere you want under the same hierarchy, either in the same "testfiles" subdirectory, or in a new "testfiles2" (or maybe "testfiles_python"?) subdirectory.  Just let me know your preference and where to get the files from.

**Comment by aerorahul** [comment]
> If you can provide the python test data, I can get the bufr.tar updated on emcrzdm (i.e. the FTP site). Currently the file contents are as follows:
> 

> So we could stash the python test data anywhere you want under the same hierarchy, either in the same "testfiles" subdirectory, or in a new "testfiles2" (or maybe "testfiles_python"?) subdirectory. Just let me know your preference and where to get the files from.

@jbathegit 
The python test data (along with the original data) is on orion at:
/work/noaa/da/rmahajan/work/JEDI/bufrWork/bufr.tar

I would like to place the python test data in `testfiles/data` as shown below:

The contents of the `bufr.tar` tarball will then be:
```
567 Orion-login-1 bufrWork $ tree /work/noaa/da/rmahajan/work/JEDI/bufrWork/bufr
bufr
└── testfiles
    ├── bufrtab.CodeFlag_LOC_0_7_1
    ├── bufrtab.CodeFlag_STD_0_13
    ├── bufrtab.TableB_LOC_0_7_1
    ├── bufrtab.TableB_STD_0_13
    ├── bufrtab.TableB_STD_0_14
    ├── bufrtab.TableB_STD_0_16
    ├── bufrtab.TableD_LOC_0_7_1
    ├── bufrtab.TableD_STD_0_13
    ├── bufrtab.TableD_STD_0_14
    ├── bufrtab.TableD_STD_0_16
    ├── data
    │   ├── 1bamua
    │   ├── gpsbufr
    │   ├── prepbufr
    │   ├── prepbufr2.ref
    │   ├── satwndbufr
    │   └── xx005
    ├── IN_1
    ├── IN_2
    ├── IN_2_bufrtab
    ├── IN_3
    ├── IN_4
    ├── IN_5
    ├── OUT_1
    ├── OUT_1_bufrtab
    ├── OUT_2
    ├── OUT_2_bufrtab
    ├── OUT_2_preAPX
    ├── OUT_3
    ├── OUT_3_bufrtab
    ├── OUT_4
    ├── OUT_4_infile1
    ├── OUT_4_infile2
    ├── OUT_5
    └── OUT_5_infile
```

**Comment by jbathegit** [comment]
Sorry, I don't have access to orion.  Could you maybe just attach the new bufr.tar to this thread so I can download it that way, or at least just a "data.tar" containing the contents of the new data subdirectory, and then I'll go ahead and take it from there?

**Comment by aerorahul** [comment]
@jbathegit 
You can download the `data/` directory from [here](https://github.com/JCSDA-internal/py-ncepbufr/tree/master/test/data) and place it in the `testfiles/` directory.

```
$> git clone https://github.com/JCSDA-internal/py-ncepbufr
$> tree -L 3 py-ncepbufr/test/data/
py-ncepbufr/test/data/
├── 1bamua
├── gpsbufr
├── prepbufr
├── prepbufr2.ref
├── satwndbufr
└── xx005
```

**Comment by jbathegit** [comment]
OK, thanks.  I'm on a separate telecon now until 12:30, but I'll try to get back to this and get it done by the time of our 2:30 meeting this afternoon.

**Comment by jbathegit** [comment]
OK, I've updated the bufr.tar on emcrzdm as requested, and you can give it a try.

**Comment by aerorahul** [comment]
@jbathegit 
Thanks!
Looks like I have some work ahead of me to figure out why the python tests are failing in CI, when they pass on my mac.

**Comment by aerorahul** [comment]
@jbathegit 
Seems like the data that you added does not match the one from the repo:
```❯❯ foreach f in $(ls -1 test/testfiles/data); do
cmp test/testfiles/data/$f ../../py-ncepbufr/test/data/$f
done
test/testfiles/data/1bamua ../../py-ncepbufr/test/data/1bamua differ: char 1, line 1
test/testfiles/data/gpsbufr ../../py-ncepbufr/test/data/gpsbufr differ: char 1, line 1
test/testfiles/data/prepbufr ../../py-ncepbufr/test/data/prepbufr differ: char 1, line 1
test/testfiles/data/prepbufr2.ref ../../py-ncepbufr/test/data/prepbufr2.ref differ: char 1, line 1
test/testfiles/data/satwndbufr ../../py-ncepbufr/test/data/satwndbufr differ: char 1, line 1
test/testfiles/data/xx005 ../../py-ncepbufr/test/data/xx005 differ: char 1, line 1
```

I replaced the files with the one from the repo and the failed tests passed.

**Comment by aerorahul** [comment]
@jbathegit 
Further, 
```
~/s/b/p/n/build ❯❯❯ ls -lrt bufr/testfiles/data 
total 48
-rw-r--r--  1 rmahajan  staff  130 Nov 12 13:31 xx005
-rw-r--r--  1 rmahajan  staff  133 Nov 12 13:31 satwndbufr
-rw-r--r--  1 rmahajan  staff  130 Nov 12 13:31 prepbufr2.ref
-rw-r--r--  1 rmahajan  staff  133 Nov 12 13:31 prepbufr
-rw-r--r--  1 rmahajan  staff  132 Nov 12 13:31 gpsbufr
-rw-r--r--  1 rmahajan  staff  133 Nov 12 13:31 1bamua
~/s/b/p/n/build ❯❯❯ ls -lrt py-ncepbufr/test/data                                
total 220576
-rw-r--r--  1 rmahajan  staff     49968 Nov  9 10:42 prepbufr2.ref
-rw-r--r--  1 rmahajan  staff     66520 Nov  9 10:42 xx005
-rw-r--r--  1 rmahajan  staff   8134808 Nov  9 10:42 gpsbufr
-rw-r--r--  1 rmahajan  staff  16495240 Nov  9 10:42 1bamua
-rw-r--r--  1 rmahajan  staff  37744912 Nov  9 10:42 satwndbufr
-rw-r--r--  1 rmahajan  staff  50422376 Nov  9 10:42 prepbufr
```

**Comment by jswhit** [comment]
The test data is stored in git lfs.  If you don't have that enabled you won't get the files when you clone the repo.

**Comment by jbathegit** [comment]
@aerorahul  I pulled these files directly by doing a clone from https://github.com/JCSDA-internal/py-ncepbufr and then cd'ing to test/data, which is what I thought you were asking me to do.  You're correct that these files are only < 200 bytes, and the contents are as follows, e.g:

% cat gpsbufr
version https://git-lfs.github.com/spec/v1
oid sha256:97fdd528d1831d345e7cf3be1dc847b68bd40825c78d09b38a0311fa2caee9cb
size 8134808

but it sounds like you want the actual file that's referenced by this link (i.e. the one which is 8134808 bytes).   I'll see if I can retrieve them for you.

**Comment by jbathegit** [comment]
OK, you can try it again now.  Just note that it increases the overall size of "bufr.tar" by a full order of magnitude, so downloads may now take longer.

**Comment by aerorahul** [comment]
@jbathegit 
That worked.  All tests are now passing.
Thanks!

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -15,13 +15,21 @@ jobs:
...
+          sudo python3 -m pip install -U numpy
+          sudo python3 -m pip install -U netCDF4
+          sudo python3 -m pip install -U pdoc
+        elif [[ ${{ matrix.os }} == "macos-10.15" ]]; then
+          brew install python3
+          pip3 install setuptools
+          pip3 install numpy
+          pip3 install netCDF4
+          pip3 install pdoc
+        fi
```
Comment (edwardhartnett) [inline .github/workflows/main.yml:34]: Excellent! I was wondering how to do this...

**Review by jbathegit — APPROVED** [review]

**Pull request merged**